Add support for Snowflake Language

primeapple opened the pull request:
This is first draft to solve #366 . It does not work yet, mainly because I had some troubles with the `QuoteTypes`.

## Review
Please review, have a look at the following things:

1. `TODO` comments show lines where I'm totally lost
2. I'm very unsure about the `reservedComments`, the `functions` and the `keywords`.
3.  Did I put the correct things in there, or do they belong somewhere else?
4. Please check the places where I used the `[thing] {a|b|c} ...` syntax, if I used them correctly.
5. Did I miss any tests?
6. Any other places that have to be changed?

## Questions
I want to support accessing elements in unstructured data.
This can be done via square brackets `[,]` or via colons `:` , like [here](https://docs.snowflake.com/en/user-guide/querying-semistructured.html#bracket-notation):
```sql
select src:salesperson.name from car_sales

select src['salesperson']['name'] from car_sales
```
How should I do this? I added `:` as an operator, is this right? How about the brackets?

Thanks for this Repo!

nene commented:
Thanks for the effort.

#### Regarding your main question about colon and square-bracket syntax.

- For brackets, adding `extraParens: ['[]']` should do the trick. See for example in `n1ql.formatter.ts`
- With colon, things are a bit trickier. You should indeed add it as an operator. However, currently the default is to format colon with a space after (as in `name: value`). To fix this, a language-specific condition should be added to `ExpressionFormatter.formatOperator()` - there already is a similar thing for `@` operator in PL/SQL.

#### In general

It would be really helpful if you could also update our [wiki](https://github.com/sql-formatter-org/sql-formatter/wiki) with information about Snowflake SQL syntax. Otherwise I'll have to do all this research on my own before I can merge this PR.

primeapple commented:
I will do all the stuff you mentioned.
Please go other the `TODO` comments in the changes. I can't run tests because I don't understand the Quote types yet...

nene commented on src/languages/snowflake/snowflake.formatter.ts:
```diff
@@ -0,0 +1,282 @@
+
+const reservedSelect = expandPhrases(['SELECT [ALL | DISTINCT]']);
+
+// https://docs.snowflake.com/en/sql-reference/sql-all.html
+//
+// 1. run in console on this page: $x('//tbody/tr/*[1]//a/span/text()').map(x => x.nodeValue)
+// 2. delete all lines that contain a sting like '(.*)', they are already covered in the list
+// 3. delete all lines that contain a sting like '<.*>', they are already covered in the list
+// 4. delete all lines that contain '…', they are part of a regex statement that can't be covered here
+// 5. Manually add 'COPY INTO'
+//
+// Steps 1-4 can be combined by the following script in the developer console:
+// $x('//tbody/tr/*[1]//a/span/text()').map(x => x.nodeValue) // Step 1
+//   filter(x => !x.match(/\(.*\)/) && !x.match(/…/) && !x.match(/<.*>/)) // Step 2-4
```
This generated list is not really particularly useful. For example `INSERT` shouldn't be listed as a separate command, instead there should be `INSERT [OVERWRITE] INTO`.

I suggest you instead look at the `reservedCommands` in `sql.formatter.ts` and use it as a base and modify it to match Snowflake syntax. Also, it would be nice to keep things in the same order (e.g. starting with `WITH`, `FROM`, `WHERE`), so it's easier to compare with other other dialect definitions in *.formatter.ts files.

primeapple replied:
Ok, I did this like in the `postgresql.formatter.ts`. I added the basic commands with the `{..} [...]` syntax, as well as the generated List because it contains many many useful phrases.

nene commented on src/languages/snowflake/snowflake.formatter.ts:
```diff
@@ -0,0 +1,282 @@
+      reservedPhrases,
+      reservedKeywords: keywords,
+      reservedFunctionNames: functions,
+      // TODO $$ is allowed, but also ' and ` how to achieve this?
+      // https://docs.snowflake.com/en/sql-reference/data-types-text.html#string-constants
+      // Backticks only allowed for inline Javascript:
+      // https://docs.snowflake.com/en/sql-reference/stored-procedures-javascript.html#line-continuation
+      stringTypes: ['$$', `'`, '`'],
```
As I understand, backticks (`` ` ``) aren't actually part of Snowflake SQL syntax (the docs suggest their usage inside JavaScript when dealing with SQL strings).

nene commented:
Regarding quote types, see: https://github.com/sql-formatter-org/sql-formatter/blob/master/src/lexer/regexFactory.ts#L99

The `-qq` and `-bs` suffixes at the end of quote types define what sort of escaping (if any) is used inside the quotes.

nene commented:
Some comments about wiki updates:

- [Parameters page](https://github.com/sql-formatter-org/sql-formatter/wiki/parameters) says that Snowflake supports `$1` parameters and links to [select docs](https://docs.snowflake.com/en/sql-reference/sql/select.html#parameters). I'm pretty sure these are something else than parameters one would use in prepared statements. I think these things are more similar to variables - they also share a very similar syntax. It's probably better to reference this syntax from [Variables page](https://github.com/sql-formatter-org/sql-formatter/wiki/variables). Probably the simplest path for supporting this syntax is also to treat them the same as variables - at lest that would be my first gut-instinct-guess.

- [Comments page](https://github.com/sql-formatter-org/sql-formatter/wiki/comments) mentions support for `# comments` but says these aren't actually supported. Syntax highlighting really shouldn't be used as a guide for what is and isn't supported. So probably best to be removed. Too bad you haven't found any official documentation for the comment syntax. Found one [Knowledge Base page](https://community.snowflake.com/s/article/how-to-comment-or-uncomment-multiple-lines-in-the-sql-worksheet) that at least mentions the `//` syntax - we could link to that.

primeapple commented:
I updated all the wiki pages for snowflake (also updated your suggestions).
One problem is the [Window-Clause](https://github.com/sql-formatter-org/sql-formatter/wiki/WINDOW-clause). Snowflake seems to not have support for a `WINDOW` clause in it's [Query Syntax)(https://docs.snowflake.com/en/sql-reference/constructs.html). There is [QUALIFY](https://docs.snowflake.com/en/sql-reference/constructs/qualify.html) but I'm unsure if this is the same?  
However, there is support for [Window Functions](https://docs.snowflake.com/en/sql-reference/functions-analytic.html) but also here I'm unsure if it is the same as in the other languages.  
Could you have a look into this?

I will rewriting the code now.

nene commented:
Thanks. Updated the docs for Snowflake window functions syntax.

nene commented:
BTW. I would appreciate if you didn't do further force-pushes to this branch. They make it hard to tell what has changed.

primeapple commented:
First refactor done.
Please the `snowflake.test` shows an infinity loop. Could you have a look into it? It happens even with a single test case.

nene commented on src/languages/snowflake/snowflake.functions.ts:
```diff
@@ -0,0 +1,614 @@
+    'BUILD_STAGE_FILE_URL',
+    'CASE',
+    'CAST',
+    '::', // this is an infix operator
```
operators shouldn't be in functions list

nene commented on src/languages/snowflake/snowflake.functions.ts:
```diff
@@ -0,0 +1,614 @@
+    'COMPLETE_TASK_GRAPHS',
+    'COMPRESS',
+    'CONCAT',
+    '||', // this is an operator like "a" || "b"
```
operators shouldn't be in functions list

nene commented on src/languages/snowflake/snowflake.functions.ts:
```diff
@@ -0,0 +1,614 @@
+    'GET_IGNORE_CASE',
+    'GET_OBJECT_REFERENCES',
+    'GET_PATH',
+    ':', // this is an infix operator AS WELL as a function
```
operators shouldn't be in functions list

nene commented on src/languages/snowflake/snowflake.functions.ts:
```diff
@@ -0,0 +1,614 @@
+    'INVOKER_ROLE',
+    'INVOKER_SHARE',
+    'IS [ NOT ] DISTINCT FROM',
+    'IS [ NOT ] NULL',
```
This `[ ... ]` syntax won't work in here because this list of strings is not wrapped inside `expandPhrases()` call.

Besides, these are again operators, not function names. That it, you won't be calling them as `IS NOT NULL(value)`

nene commented on src/languages/snowflake/snowflake.functions.ts:
```diff
@@ -0,0 +1,614 @@
+    'SYSTEM$VERIFY_EXTERNAL_OAUTH_TOKEN',
+    'SYSTEM$WAIT',
+    'SYSTEM$WHITELIST',
+    'SYSTEM$WHITELIST_PRIVATELINK',
```
Currently these won't work because the functions list gets concatenated to a regex and `$` will attempt to match end-of-line.

I should fix this by adding escaping of function names before constructing a regex.

nene reviewed (commented):
The reason for the infinite loop is `||` operator in functions list, which gets concatenated into a long regex containing all function names and ends up matching empty string.

nene reviewed (commented):
Some additional problems that popped up to me

nene commented:
Regarding the formatting of `:` operator. I made a change to make the customization of formatting simpler: #466

One should now be able to simply add this to SnowflakeFormatter class:

```ts
formatOptions() {
  return { alwaysDenseOperators: [':'] };
}
```

nene commented:
Also added escaping of keywords, so the `$` character in function names should no more pose a problem: #473

primeapple commented on src/languages/snowflake/snowflake.formatter.ts:
```diff
@@ -0,0 +1,349 @@
+      stringTypes: ['$$', `''-qq`],
+      identTypes: ['""-qq'],
+      variableTypes: [
+        // for accessing columns at certain positons in the table
+        { regex: '$[1-9]d*' },
+        // identifier style syntax
+        { regex: '$([_a-zA-Z][_a-zA-Z0-9$]*)' },
```
Changed `d*` to `\\d*` to support any number sequence

primeapple commented:
I'm sorry, I screwed up the commits by including your commits into this PR. This is why I usually force push, it allows me to cleanly integrate commits to master with `rebase` to the feature branch and then force push the feature branch.

primeapple commented:
Ok, I reverted your commits in this PR and instead merged master into this branch. Everything fine now. Puhhh...

The only new important commit for you is [8dc6f51](https://github.com/sql-formatter-org/sql-formatter/pull/454/commits/8dc6f518a2e0bf9d76458c6a24274795b14f5a3b)

primeapple commented on test/snowflake.test.ts:
```diff
@@ -0,0 +1,81 @@
+    `);
+  });
+
+  it(`formats ':' path-operator without spaces`, () => {
+    expect(format(`SELECT foo:bar FROM test`)).toBe(dedent`
+      SELECT
+        foo:bar
+      FROM
+        test
+    `);
+  });
+
+  it(`does not support ':' path-operator with spaces`, () => {
+    expect(format(`SELECT foo : bar FROM test`)).toThrowError('Parse error');
+  });
```
These tests fail, unsure why.

nene replied:
Ah yeah, I forgot that there's a special formatting rule about the `:` operator and so the `alwaysDenseOperators` won't work with `:`. I'll have to fix that.

However, the second test case surprising for me. Is it really the case that Snowflake does not allow spaces around `:` operator? In that case we'll have to treat it as part of identifier instead, otherwise extra whitespace always gets allowed.

primeapple replied:
I'll investigate this.

primeapple replied:
Wow, I did not know this. It does accept `:` with spaces around!

primeapple replied:
I fixed these tests and added some for the bracket and dot notation as well. Now the only thing failing is the `:` operator since it gets succeeding whitespace.

primeapple commented on src/languages/snowflake/snowflake.formatter.ts:
```diff
@@ -0,0 +1,352 @@
+  'FETCH [FIRST | NEXT]',
+  // Data manipulation
+  // - insert:
+  'INSERT [OVERWRITE] [ALL INTO | INTO | ALL | FIRST]',
```
Are multiple following `[...]` blocks supported? I'm unsure because of the spaces in between, I dont want it to become something like `INSERT  INTO` (2 spaces in between).

nene replied:
Yep that's fine. The spaces will be taken care of by `expandPhrases()`. In the end though any number of whitespace characters between these words will be allowed by the parser.

primeapple commented:
1. Is there an option to define how identifiers are allowed to look like?
In Snowflake they can include any symbol when inside `""`, but without quoting they have a certain rule. There is the `identChars` Config option, but I'm unsure if this refers to quoted or unquoted identifiers.

2. Is there an option to test variables?

3. Did I miss any obvious `support...` tests?

primeapple commented on src/languages/snowflake/snowflake.formatter.ts:
```diff
@@ -0,0 +1,352 @@
+      reservedFunctionNames: functions,
+      stringTypes: ['$$', `''-qq-bs`],
+      identTypes: ['""-qq'],
+      variableTypes: [
+        // for accessing columns at certain positons in the table
+        { regex: '[$][1-9]\\d*' },
+        // identifier style syntax
+        { regex: '[$][_a-zA-Z][_a-zA-Z0-9$]*' },
+      ],
```
Are these correctly written?

1. Should be anything like `$number`, where `number > 0`
2. Shoud be anything like `$identifier`, where identifier starts with char `a-Z`, can contain any combination of multiple `$`, `a-Z` and `0-9`

nene replied:
Looks fine to me. I'd suggest simplifying the first one to just:

```
[$][0-9]+
```

It's OK if we accept SQL that's not fully correct - the goal is formatting, not validation. It would only pose a problem with `$0` would be some sort of special syntax that we'd like to handle elsewhere.

primeapple replied:
I see your point, but why not just leave it? It's not that complicated, I can rewrite it to 
`[$][1-9][0-9]*`
if that's better readable?

nene replied:
Might as well just leave it as is.

nene commented:
> Is there an option to define how identifiers are allowed to look like?

Yep, there are two options for that:

- `identTypes` describes quoted identifiers (with all characters allowed inside quotes).
- `identChars` describes unquoted identifiers.

Both of these seem to be configured correctly.

> Is there an option to test variables?

No, the variable syntax differs quite a bit between SQL dialects. Have a look at the variable tests in `behavesLikeMariaDbFormatter.ts` and create some custom tests for Snowflake.

> Did I miss any obvious support... tests?

Doesn't seem so at first glance.

nene commented on test/features/truncateTable.ts:
```diff
@@ -18,6 +19,16 @@ export default function supportsTruncateTable(
+      expect(result).toBe(dedent`
+        TRUNCATE TABLE IF EXISTS
+          Customers;
+`);
```
Though this test too is not that useful. This syntax is only supported by Snowflake, and like elsewhere, testing all these syntax variations doesn't really add much value.

nene replied:
Let's also fix this inconsistent indentation in here.

nene commented on test/snowflake.test.ts:
```diff
@@ -0,0 +1,81 @@
+  supportsSetOperations(format, ['UNION', 'UNION ALL', 'MINUS', 'EXCEPT', 'INTERSECT']);
+  supportsLimiting(format, { limit: true, offset: true, fetchFirst: true, fetchNext: true });
+
+  it('allows $ character as part of identifiers if inclosed by "', () => {
+    expect(format('SELECT "foo$"')).toBe(dedent`
+      SELECT
+        "foo$"
+    `);
+  });
```
We should instead test that $ character can be used inside unquoted identifiers. The quoted ones are already covered by `supportsIdentifiers()` test.

nene commented on src/languages/snowflake/snowflake.formatter.ts:
```diff
@@ -0,0 +1,352 @@
+  // Data manipulation
+  // - insert:
+  'INSERT [OVERWRITE] [ALL INTO | INTO | ALL | FIRST]',
+  '[THEN | WHEN] INTO',
```
This is not correct. As I understand, the syntax is:

```
WHEN condition THEN INTO tablename
ELSE INTO tablename
```

We could break this up into the following:

```
'WHEN',
'THEN INTO',
'ELSE',
```

primeapple replied:
Ahh, you are right. Good find.
We can just remove `WHEN` and write `[THEN | ELSE] INTO`

primeapple commented on src/languages/snowflake/snowflake.formatter.ts:
```diff
@@ -0,0 +1,352 @@
+  '{CROSS | NATURAL} JOIN',
+]);
+
+const reservedPhrases = expandPhrases(['{ROWS | RANGE} BETWEEN']);
```
Snowflake does not support these.

nene replied:
Does it really not? I see them in [snowflake documentation.](https://docs.snowflake.com/en/sql-reference/sql/create-table-constraint.html)

Plus, to test these, we should include `supportsConstraints()` helper.

primeapple commented on test/features/schema.ts:
```diff
@@ -2,11 +2,11 @@ import dedent from 'dedent-js';
-export default function supportsSchema(format: FormatFn) {
-  it('formats simple SET SCHEMA statements', () => {
-    const result = format('SET SCHEMA schema1;');
+export default function supportsSchema(format: FormatFn, setKeyword = 'SET') {
```
You're right, restored this.

primeapple commented on test/features/createView.ts:
```diff
@@ -56,4 +57,13 @@ export default function supportsCreateView(
   }
+
+  if (ifNotExists) {
+    it('formats short CREATE TABLE IF NOT EXISTS', () => {
```
Do you want me to open an issue for this?

nene replied:
No worries, I mainly just wrote this comment as a not for myself.

primeapple commented:
I understand your concerns regarding the tests.
However in my opinion they have the purpose of explaining the accepted syntax without having to look into the actual source files.

For example with `supportsTruncateTable(format, { withoutTable: true, ifExists: true })` we can very easily see all the dialects that support `TRUNCATE IF EXISTS`, without having to manually look into the source files?

We've 3 options:
1. Leave the tests in this PR, add the other languages that support them later (I can do this if you want?)
2. Add the tests completly in a later PR
3. Remove the tests

Option 3. would decrease test coverage and verbosity for the sake of a few less lines and some milisecons of test time.

You are the maintainer, it's up to you.

nene commented:
OK, let the tests be for now as they are. My main concern was with the `SET SCHEMA` test. Others are not that big of a deal.

nene commented:
OK. I fixed the `:` handling. If you now merge in the `master`, it should work.

Though now you'll also need to add the `::` operator to the `alwaysDenseOperators` config.

primeapple commented:
Everything done.
Is this mergable now? :smiley:

nene commented on src/languages/snowflake/snowflake.formatter.ts:
```diff
@@ -0,0 +1,352 @@
+  // Data manipulation
+  // - insert:
+  'INSERT [OVERWRITE] [ALL INTO | INTO | ALL | FIRST]',
+  '[THEN | ELSE] INTO',
```
Should instead use:
```suggestion
  '{THEN | ELSE} INTO',
```

nene commented on test/snowflake.test.ts:
```diff
@@ -0,0 +1,87 @@
+          foo:bar.baz
+      `);
+    });
+    it(`formats brackets notation without spaces`, () => {
+      expect(format(`SELECT foo [ 'bar' ]`)).toBe(dedent`
+        SELECT
+          foo['bar']
+      `);
+    });
```
No need for this test, it's already covered by `supportsArrayAndMapAccessors()`

nene reviewed (commented):
Found a few more issues.

primeapple commented:
I somehow can't answer your comments, so I'll do it here:

I added to reservedPhrases:
```js
  'MATCH {FULL | SIMPLE | PARTIAL}',
  'ON {UPDATE | DELETE} {CASCADE | SET NULL | SET DEFAULT | RESTRICT | NO ACTION}',
  'INITIALLY {DEFERRED | IMMEDIATE}',
```
This breaks the `supportConstraints` tests, because we are doing `ON UPDATE CURRENT_TIMESTAMP` there.
This test is problematic anyway because we are using it for many languages that don`t even support this line (Postgresql for example).

I fixed the test, we can add the correct specifications for other languages later.

nene commented:
Thanks for putting up with all my complaints :)

I'll merge this in and create a beta release out of it.

primeapple commented:
Thank you for all the fast feedback. It was my first open source contribution and you made the process very easy.

nene commented:
It's now released as 11.0.0-beta.1
